core/dagop.go

```diff
@@ -321,44 +339,384 @@ func (op RawDagOp) Exec(ctx context.Context, g bksession.Group, inputs []solver.
+			mount.Readonly = true
+		}
+
+		if mnt.CacheVolumeID != "" {
+			mount.Output = pb.SkipOutput
+			mount.MountType = pb.MountType_CACHE
+			mount.CacheOpt = &pb.CacheOpt{
+				ID: mnt.CacheVolumeID,
+			}
+			switch mnt.CacheSharingMode {
+			case CacheSharingModeShared:
+				mount.CacheOpt.Sharing = pb.CacheSharingOpt_SHARED
+			case CacheSharingModePrivate:
+				mount.CacheOpt.Sharing = pb.CacheSharingOpt_PRIVATE
+			case CacheSharingModeLocked:
+				mount.CacheOpt.Sharing = pb.CacheSharingOpt_LOCKED
+			}
+		}
+
+		if mnt.Tmpfs {
+			mount.Output = pb.SkipOutput
+			mount.MountType = pb.MountType_TMPFS
+			mount.TmpfsOpt = &pb.TmpfsOpt{
+				Size_: int64(mnt.Size),
+			}
+		}
+
+		mounts = append(mounts, mount)
+		if mount.Output != pb.SkipOutput {
+			outputIdx++
+		}
+
+		return nil
+	}
+
+	// handle our normal mounts
+	if err := addMount(ContainerMount{Source: container.FS, Target: "/"}); err != nil {
+		return nil, nil, err
+	}
+	if err := addMount(ContainerMount{Source: container.Meta, Target: buildkit.MetaMountDestPath, SourcePath: buildkit.MetaMountDestPath}); err != nil {
+		return nil, nil, err
+	}
+	for _, mount := range container.Mounts {
+		if err := addMount(mount); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	// handle secret mounts
+	for _, secret := range container.Secrets {
+		if secret.MountPath == "" {
+			continue
+		}
+		uid, gid := 0, 0
+		if secret.Owner != nil {
+			uid, gid = secret.Owner.UID, secret.Owner.GID
+		}
+		mount := &pb.Mount{
+			Input:     pb.Empty,
+			Dest:      secret.MountPath,
+			MountType: pb.MountType_SECRET,
+			SecretOpt: &pb.SecretOpt{
+				ID:   secret.Secret.ID().Digest().String(),
+				Uid:  uint32(uid),
+				Gid:  uint32(gid),
+				Mode: uint32(secret.Mode),
+			},
+		}
+		mounts = append(mounts, mount)
+	}
+
+	// handle socket mounts
+	for _, socket := range container.Sockets {
+		if socket.ContainerPath == "" {
+			return nil, nil, fmt.Errorf("unsupported socket: only unix paths are implemented")
+		}
+		uid, gid := 0, 0
+		if socket.Owner != nil {
+			uid, gid = socket.Owner.UID, socket.Owner.GID
+		}
+		mount := &pb.Mount{
+			Input:     pb.Empty,
+			Dest:      socket.ContainerPath,
+			MountType: pb.MountType_SSH,
+			SSHOpt: &pb.SSHOpt{
+				ID:   socket.Source.LLBID(),
+				Uid:  uint32(uid),
+				Gid:  uint32(gid),
+				Mode: 0o600, // preserve default
+			},
+		}
+		mounts = append(mounts, mount)
+	}
+
+	return mounts, states, nil
+}
+
+// setAllContainerMounts is the reverse of getAllContainerMounts, and rewrites
+// the container mounts based on the mounts and states specified.
+func setAllContainerMounts(ctx context.Context, container *Container, mounts []*pb.Mount, states []llb.State) error {
+	for _, mount := range mounts {
+		if mount.Output == -1 {
+			continue
+		}
+
+		st := states[mount.Output]
+
+		def, err := st.Marshal(ctx, llb.Platform(container.Platform.Spec()))
+		if err != nil {
+			return err
+		}
+
+		switch mount.Output {
+		case -1:
```

Comment:
would this result in us skipping re-adding all the mounts like cache volume, tmpfs, secret, etc. ? or am I missing something where that's handled?